### PR TITLE
WIP: Generalized perspective support

### PR DESCRIPTION
**Pull request by ericlux**
## Description
Support for generic handling of the perspective transform has been a longstanding issue dating back to pre 1.0 days. We (@phschiele  and I) use the fact that the perspective transform of a function with a given graph form is trivially obtained. We implement this functionality as a new atom called `perspective` which takes in a (convex) scalar expression `f ` and a nonnegative affine scalar `s`. Under the hood, the canonicalization computes `f`'s graph form, adds the new variable `s`, and represents the perspective of `f` by returning the epigraph variable `t`.

For further reference on the graph form perspective rule, see @moehle's [paper](https://www-sciencedirect-com.stanford.idm.oclc.org/science/article/pii/S0167691115001747).

This is still a work in progress and we hope to get some feedback on the general implementation idea. We provide extensive tests for problems involving the second order cone, the exponential cone, the power cone, and the psd cone. There is still some minor additions: dealing with concave functions, grad, elementwise capability, but we think the core functionality is there.

Issue link (if applicable): #420

## Type of change
- [x] New feature (backwards compatible)
- [ ] New feature (breaking API changes)
- [ ] Bug fix
- [ ] Other (Documentation, CI, ...)

## [Contribution checklist](https://www.cvxpy.org/contributing/index.html#contribution-checklist)
- [x] Add our license to new files.
- [x] Check that your code adheres to our coding style.
- [x] Write unittests.
- [x] Run the unittests and check that they’re passing.
- [x] Run the benchmarks to make sure your change doesn’t introduce a regression.

### CODE REVIEW

**Comment by adishavit** [comment]
The link to the paper requires a password.  
Can you provide an alternate?  
Is [this](https://web.stanford.edu/~boyd/papers/sw_aff_ctrl.html) the one?

**Comment by ericlux** [comment]
Yes, sorry about that! That’s the correct paper.

File: cvxpy/atoms/elementwise/elementwise.py [hunk 1]
```diff
@@ -60,7 +60,7 @@ def elemwise_grad_to_diag(value, rows, cols):
         """
         if not np.isscalar(value):
             value = value.ravel(order='F')
-        return sp.dia_matrix((value, [0]), shape=(rows, cols)).tocsc()
+        return sp.dia_matrix(([value], [0]), shape=(rows, cols)).tocsc()
```
Comment (SteveDiamond) [inline cvxpy/atoms/elementwise/elementwise.py:63]: why is this changed?

File: cvxpy/atoms/elementwise/exploration.ipynb [hunk 2]
```diff
@@ -0,0 +1,72 @@
+{
```
Comment (SteveDiamond) [inline cvxpy/atoms/elementwise/exploration.ipynb:1]: Is this file intended to be committed?

File: cvxpy/atoms/elementwise/elementwise.py (hunk 1 above)
Comment (ericlux) [reply cvxpy/atoms/elementwise/elementwise.py:63]: This fails under scipy1.9, nothing to do with the perspective. Philipp and I just ID'd it along the way.

File: cvxpy/atoms/elementwise/exploration.ipynb (hunk 2 above)
Comment (ericlux) [reply cvxpy/atoms/elementwise/exploration.ipynb:1]: Nope, I'm bad at git! Clearing now

File: cvxpy/reductions/dcp2cone/atom_canonicalizers/perspective_canon.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    constraints.append(-c@x_canon + t - s*d >= 0)
+
+    # recover initial variables
+
+    end_inds = sorted(prob_canon.var_id_to_col.values()) + [x_canon.shape[0]]
+
+    for var in expr.f.variables():
+        start_ind = prob_canon.var_id_to_col[var.id]
+        end_ind = end_inds[end_inds.index(start_ind)+1]
+        if var.is_psd():
```
Comment (SteveDiamond) [inline cvxpy/reductions/dcp2cone/atom_canonicalizers/perspective_canon.py:80]: Should be var.is_symmetric()?

Comment (ericlux) [reply cvxpy/reductions/dcp2cone/atom_canonicalizers/perspective_canon.py:80]: I had this before but ran into a failed test because a scalar variable is symmetric. But thanks for catching this; the matrix totally could be NSD also. Any thoughts on how to handle this? In general wasn’t sure the best generic way to set these “outer” variables (that the user constructing the perspective created) equal to their underlying counterpart which is exposed in the canonicalization (which causes this outer matrix to engage in an equality constraint with a different shape variable in x_canon).

File: cvxpy/tests/test_perspective.py [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+
+
+def test_rel_entr():
+    x = cp.Variable()
+    s = cp.Variable(nonneg=True)
+    f = -cp.log(x)
+    obj = cp.perspective(f, s)
+    constraints = [1 <= s, s <= 2, 1 <= x, x <= 2]
+    prob = cp.Problem(cp.Minimize(obj), constraints)
+    prob.solve(solver=cp.MOSEK)
```
[the remark this answers is not in the document]
Comment (ericlux) [reply cvxpy/tests/test_perspective.py]: Moved to default solvers. Resolved by [b9f1fc4](https://github.com/cvxpy/cvxpy/pull/1867/commits/b9f1fc4a1244738ad3f7d8bbd3db4de90f87657d)

**Comment by SteveDiamond** [comment]
@ericlux we had a discussion about this PR, and it's almost ready to be integrated for 1.3. The only remaining TODO is to add an assert that no parameters are involved, since that case isn't handled.

**Comment by ericlux** [comment]
Oh awesome! In that case we will defer parameters to later, and add the assert!

**Comment by ericlux** [comment]
@SteveDiamond updated the documentation as discussed
Screenshot 2022-11-19 at 2 11 59 PM

**Comment by ericlux** [comment]
@SteveDiamond tests passing after diffcp fix; should be good to go.

**Review by SteveDiamond — APPROVED** [review]
LGTM. Could you clarify how parameters can be used with parameters? It looks like there are no restrictions.

**Comment by ericlux** [comment]
We added `chain = aux_prob._construct_chain(solver_opts=solver_opts, ignore_dpp=True)` in the canon method. So any parameter will just be evaluated without getting the dpp benefits for repeated solves.

**Comment by SteveDiamond** [comment]
> We added `chain = aux_prob._construct_chain(solver_opts=solver_opts, ignore_dpp=True)` in the canon method. So any parameter will just be evaluated without getting the dpp benefits for repeated solves. We'll double check to make sure is_dpp() returns false for a parameterized perspective expression.

Got it, thanks! If the PR looks good to @phschiele we can merge then. Thanks for all your hard work on this!
